### DISCUSSION TITLE
Inconsistent State Change Rendering in Component

### DISCUSSION THREAD

**Discussion by guidodf98** [livewire/livewire#7186] [1. Help] [answered]
### Problem:
I'm encountering an issue with a component in my application that fetches and displays appointments from a database. The component allows me to change the state of these appointments using a button. The problem I'm facing is that when I change the state of an appointment by clicking the button, the change is not immediately reflected in the view. It only updates after the second click.

### Code Sample:
Here's a snippet of my component code:

      <x-section class="p-5 mb-4">
        <ol class="mt-3 divide-y divider-gray-200 dark:divide-gray-700">
          @foreach ($turnos as $turno)
            <li class="items-center block p-3 hover:bg-gray-100 dark:hover:bg-gray-700 justify-between {{ $turno->estado == 'finalizado' ? 'turno-finalizado flex' : 'sm:flex' }} {{ $turno->estado == 'ausente' ? 'bg-opacity-60 bg-gray-200 dark:bg-gray-900' : '' }}" wire:key='{{ $turno->id }}'>
              <div class="xs:flex">
                <div class="text-gray-600 dark:text-gray-400">
                  <div class="text-base font-normal">
                    <span class="font-medium text-gray-900 dark:text-white">{{ $turno->hora }}</span>
                    -
                    <span class="font-medium text-gray-900 dark:text-white">{{ $turno->paciente->user->nombre }} {{ $turno->paciente->user->apellido }}</span>
                  </div>
                  <span class="uppercase inline-flex items-center font-normal text-gray-500 dark:text-gray-400">
                    {{ $turno->estado }}
                  </span>
                </div>
              </div>
              <div class="pt-2 {{ $turno->estado != 'finalizado' ? 'flex justify-center gap-5 sm:gap-2' : '' }}">
                @if ($turno->estado == 'atendiendo')
                  <button wire:click="cambiarEstado('finalizado',{{ $turno->id }})">
                    Finalizar
                  </button>
                @elseif ($turno->estado == 'finalizado')
                  <x-svgs.turno-finalizado />
                  <x-svgs.turno-finalizado-volver turnoId="{{ $turno->id }}" />
                @elseif ($turno->estado == 'ausente')
                  <x-button wire:click="cambiarEstado('atendiendo',{{ $turno->id }})">Atender</x-button>
                @else
                  <x-button wire:click="cambiarEstado('atendiendo',{{ $turno->id }})">Atender</x-button>
                  <x-button wire:click="cambiarEstado('ausente',{{ $turno->id }})">Ausente</x-button>
                @endif

              </div>
            </li>
          @endforeach
        </ol>
      </x-section>

And here's the related class code:

```
<?php

namespace App\Livewire\Vistas\Medico;

use App\Models\Medico;
use App\Models\Turno;
use Carbon\Carbon;
use Illuminate\Support\Facades\Auth;
use Livewire\Attributes\On;
use Livewire\Component;
use Livewire\WithPagination;

class Turnos extends Component {
  use WithPagination;
  public $turnos = [];
  public $fechaFormateada;

  public function mount() {
    $this->cargarTurnosDeFecha(date('m/d/Y'));
  }

  public function render() {
    return view('livewire.vistas.medico.turnos');
  }

  public function cambiarEstado($estado, Turno $turno) {
    $turno->estado = $estado;
    $turno->save();
  }

  #[On('cargarTurnosDeFecha')]
  public function cargarTurnosDeFecha($fecha) {
    $medico = Medico::where('user_id', Auth::user()->id)->first();
    $fecha = Carbon::createFromFormat('m/d/Y', $fecha)->format('Y-m-d');

    $this->turnos = Turno::whereNotNull('paciente_id')
      ->where('medico_id', $medico->id)
      ->where('dia', $fecha)
      ->get();
  }
}
```

### Steps to Reproduce:

1. Load the component with appointment data.
2. Click on the "Finalizar" button for an appointment to change its state.
3. Observe that the state change is not immediately reflected in the view.
4. The view updates only after a second click on the same button.

### Expected Behavior:
I expect the view to update immediately when I change the state of an appointment by clicking the button.

### Additional Information:
I've confirmed that the render method is being called after changing the state, but the view is not updating as expected.

### Environment:
- Laravel: 10.10
- Livewire: 3.0
- PHP: 8.1

**Comment by joshhanley** [comment, marked as answer]
@guidodf98 the issue is that Livewire has already loaded your Eloquent Collection before the change has been applied. So you need to re-get the models from the database. Or a better solution, is don't assign Eloquent Collections to public properties. Instead make use of a computed property for Eloquent Collections. https://livewire.laravel.com/docs/computed-properties

Hope this helps!

**Reply by guidodf98** [reply]
Thanks Josh for the solution! I followed your advice and used computed properties. Much appreciated!
